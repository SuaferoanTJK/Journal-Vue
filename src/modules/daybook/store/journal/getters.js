// export const getter = (state) => {
//     return state;
// }

export const getEntriesByTerm =
  (state) =>
  (term = "") => {
    if (term.length === 0) return state.entries;
    return state.entries.filter((entry) =>
      entry.text.toLowerCase().includes(term.toLocaleLowerCase())
    );
  };
export const getEntryByID = (state) => (id) => {
  const entry = state.entries.find((entry) => entry.id == id);
  if (!entry) return;
  return { ...entry };
};
