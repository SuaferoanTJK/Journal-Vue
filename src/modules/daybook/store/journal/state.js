export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(),
      date: new Date().toDateString(),
      text: "Et exercitation mollit aute cupidatat nostrud culpa occaecat cupidatat dolor enim minim esse. Aute do magna ut consectetur eu mollit. Nostrud proident excepteur esse elit nisi. Occaecat exercitation aliquip laborum cupidatat. Sunt dolor est laboris veniam ipsum. Deserunt elit commodo labore est in cupidatat et dolore.",
      picture: null,
    },
    {
      id: new Date().getTime() + 1000,
      date: new Date().toDateString(),
      text: "Do aute fugiat do qui nostrud nulla qui elit est duis laboris labore ipsum do. Eiusmod quis consectetur reprehenderit aute voluptate elit dolor. Consectetur exercitation duis velit deserunt ipsum amet ad et consequat ut.",
      picture: null,
    },
    {
      id: new Date().getTime() + 2000,
      date: new Date().toDateString(),
      text: "Elit amet Lorem aliquip voluptate deserunt mollit anim duis. Eiusmod voluptate amet et reprehenderit consequat nostrud cupidatat. Magna deserunt consequat est proident exercitation eiusmod qui. Do ad ut proident sint labore ipsum mollit minim sint. Pariatur minim culpa aliquip amet amet id duis id culpa Lorem irure Lorem qui ut. Culpa ea eiusmod consequat exercitation et incididunt velit.",
      picture: null,
    },
  ],
});
