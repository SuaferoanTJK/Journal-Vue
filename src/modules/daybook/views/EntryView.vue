<script>
import Swal from "sweetalert2";
import { mapGetters, mapActions } from "vuex";
import Fab from "../components/Fab";
import { formatDate } from "../helpers/formatDate";
import uploadImage from "../helpers/uploadImage";

export default {
  name: "EntryView",
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      entry: null,
      localImage: null,
      file: null,
    };
  },
  components: {
    Fab,
  },
  computed: {
    ...mapGetters("journal", ["getEntryByID"]),
    date() {
      const date = formatDate(this.entry.date);
      return date;
    },
  },
  methods: {
    ...mapActions("journal", ["createEntry", "updateEntry", "deleteEntry"]),
    loadEntry() {
      let entry;
      if (this.id === "new") {
        entry = {
          text: "",
          date: new Date().getTime(),
        };
      } else {
        entry = this.getEntryByID(this.id);
        if (!entry) return this.$router.push({ name: "noEntry" });
      }
      this.entry = entry;
    },
    async saveEntry() {
      Swal.fire({
        title: "Please wait",
        allowOutsideClick: false,
      });
      Swal.showLoading();

      const picture = await uploadImage(this.file);
      this.entry.picture = picture;

      if (this.entry.id) {
        await this.updateEntry(this.entry);
      } else {
        const id = await this.createEntry(this.entry);
        this.$router.push({ name: "entry", params: { id } });
      }
      this.file = null;
      Swal.fire("Saved", "Entry registered successfully", "success");
    },
    async onDeleteEntry() {
      const { isConfirmed } = await Swal.fire({
        title: "Are you sure?",
        text: "Once deleted, it can't be recovered",
        showDenyButton: true,
        confirmButtonText: "Yes, I'm sure",
      });
      if (isConfirmed) {
        Swal.fire({
          title: "Please wait",
          allowOutsideClick: false,
        });
        Swal.showLoading();
        await this.deleteEntry(this.entry.id);
        this.$router.push({ name: "noEntry" });
        Swal.fire("Deleted", "", "success");
      }
    },
    onSelectedImage(event) {
      const file = event.target.files[0];
      if (!file) {
        this.localImage = null;
        this.file = null;
        return;
      }
      this.file = file;
      const fr = new FileReader();
      fr.onload = () => (this.localImage = fr.result);
      fr.readAsDataURL(file);
    },
    onSelectImage() {
      this.$refs.imageSelector.click();
    },
  },
  created() {
    this.loadEntry();
  },
  watch: {
    id() {
      this.loadEntry();
    },
  },
};
</script>

<template>
  <div
    v-if="entry"
    class="shadow-lg md:min-h-[85vh] p-5 border-2 rounded relative"
  >
    <div class="flex flex-col lg:flex-row gap-2 justify-between">
      <h1 class="font-bold text-3xl text-green-800">
        <p>{{ date }}</p>
      </h1>
      <div class="flex flex-col gap-4">
        <input
          type="file"
          @change="onSelectedImage"
          ref="imageSelector"
          v-show="false"
          accept="image/png, image/jpeg, image/jpg"
        />
        <div class="flex gap-2 justify-between">
          <button
            v-if="entry.id"
            @click="onDeleteEntry"
            class="deleteBtn w-full h-12 items-center text-sm bg-red-600 hover:bg-red-700 py-1 px-3 text-white font-semibold rounded"
          >
            Delete
          </button>
          <button
            @click="onSelectImage"
            class="w-full h-12 items-center text-sm bg-lime-600 hover:bg-lime-700 py-1 px-3 text-white font-semibold rounded"
          >
            Upload photo
          </button>
        </div>
      </div>
    </div>
    <div class="mt-3">
      <div>
        <label for="text" class="font-semibold">Text</label>
        <textarea
          id="text"
          placeholder="What happened today?"
          v-model="entry.text"
          class="rounded w-full py-2 px-3 text-gray-700 leading-tight min-h-[20vh] md:min-h-[50vh]"
        ></textarea>
      </div>
      <div v-if="localImage">
        <label class="font-semibold">Image</label>
        <img :src="localImage" alt="entry-picture" class="w-80 md:w-60" />
      </div>
      <div v-if="entry.picture && !localImage">
        <label class="font-semibold">Image</label>
        <img :src="entry.picture" alt="entry-picture" class="w-80 md:w-60" />
      </div>
      <Fab
        @on:click="saveEntry"
        icon="fa-save"
        class="absolute bottom-3 right-3"
      />
    </div>
  </div>
</template>
