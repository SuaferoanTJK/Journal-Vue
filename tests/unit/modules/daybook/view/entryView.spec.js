import { createStore } from "vuex";
import { shallowMount } from "@vue/test-utils";
import Swal from "sweetalert2";
import EntryView from "@/modules/daybook/views/EntryView.vue";
import journal from "@/modules/daybook/store";
import { journalState } from "../../../mock-data/test-journal-state";

const createVuexStore = (initialState) =>
  createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...initialState },
      },
    },
  });

jest.mock("sweetalert2", () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
}));

describe("EntryView Unit Testing", () => {
  const store = createVuexStore(journalState);
  store.dispatch = jest.fn();

  const mockRouter = {
    push: jest.fn(),
  };

  let wrapper;
  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryView, {
      props: {
        id: "-N8ehquIXWnUzhMQqxVe",
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    });
  });

  test("Should redirect because the ID doesn't exist", () => {
    shallowMount(EntryView, {
      props: {
        id: "ID that doesn't exist",
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    });
    expect(mockRouter.push).toHaveBeenCalledWith({ name: "noEntry" });
  });
  test("Should display entry", () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(mockRouter.push).not.toHaveBeenCalled();
  });
  test("Should delete the entry and exit", async () => {
    Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }));
    await wrapper.find(".deleteBtn").trigger("click");
    expect(Swal.fire).toHaveBeenCalledWith({
      title: "Are you sure?",
      text: "Once deleted, it can't be recovered",
      showDenyButton: true,
      confirmButtonText: "Yes, I'm sure",
    });
  });
});
