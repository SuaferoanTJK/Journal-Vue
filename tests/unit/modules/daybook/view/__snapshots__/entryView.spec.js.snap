// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryView Unit Testing Should display entry 1`] = `
<div class="shadow-lg md:min-h-[85vh] p-5 border-2 rounded relative">
  <div class="flex flex-col lg:flex-row gap-2 justify-between">
    <h1 class="font-bold text-3xl text-green-800">
      <p>Thursday, August 04, 2022</p>
    </h1>
    <div class="flex flex-col gap-4"><input type="file" accept="image/png, image/jpeg, image/jpg" style="display: none;">
      <div class="flex gap-2 justify-between"><button class="deleteBtn w-full h-12 items-center text-sm bg-red-600 hover:bg-red-700 py-1 px-3 text-white font-semibold rounded"> Delete </button><button class="w-full h-12 items-center text-sm bg-lime-600 hover:bg-lime-700 py-1 px-3 text-white font-semibold rounded"> Upload photo </button></div>
    </div>
  </div>
  <div class="mt-3">
    <div><label for="text" class="font-semibold">Text</label><textarea id="text" placeholder="What happened today?" class="rounded w-full py-2 px-3 text-gray-700 leading-tight min-h-[20vh] md:min-h-[50vh]"></textarea></div>
    <!--v-if-->
    <!--v-if-->
    <fab-stub icon="fa-save" class="absolute bottom-3 right-3"></fab-stub>
  </div>
</div>
`;
